src/codegen/optimizer.lisp

```diff
@@ -316,6 +320,106 @@
+           (type (integer 0)    max-depth)
+           (type (integer 0)    max-unroll)
+           (type function       heuristic)
+           (values node &optional))
+  (flet ((try-inline (node call-stack)
+           (declare (type (or node-application node-direct-application) node)
+                    (type parser:identifier-list                        call-stack)
+                    (values (or node-let node-application node-direct-application) &optional))
+           (or
+            (alexandria:when-let*
+                ((name (node-rator-name node))
+                 (code (tc:lookup-code env name :no-error t))
+                 (_    (and (node-abstraction-p code)
+                            (funcall heuristic code)
+                            (<= (length call-stack)
+                                max-depth)
+                            (<= (count name call-stack)
+                                max-unroll)
+                            (= (length (node-abstraction-vars code))
+                               (length (node-rands node))))))
+              (multiple-value-bind (bindings subs)
+                  (loop :for var :in (node-abstraction-vars code)
+                        :for val :in (node-rands node)
+                        :for new-var := (gentemp (symbol-name var))
+                        :collect (cons new-var val) :into bindings
+                        :collect (make-ast-substitution
+                                  :from var
+                                  :to (make-node-variable
+                                       :type (node-type val)
+                                       :value new-var))
+                          :into subs
+                        :finally (return (values bindings subs)))
+                (make-node-let
+                 :type     (node-type node)
+                 :bindings bindings
+                 :subexpr  (funcall *traverse*
+                                    (rename-type-variables (apply-ast-substitution
+                                                            subs
+                                                            (node-abstraction-subexpr code)))
+                                    (cons name call-stack)))))
+            node))
+         (inline-let-without-breaking-recursive-cons (node call-stack)
+           (declare (type node-let               node)
+                    (type parser:identifier-list call-stack)
+                    (values node-let &optional))
+           (make-node-let
+            :type     (node-type node)
+            :bindings (loop :for (name . subnode) :in (node-let-bindings node)
+                            :collect (cons name
+                                           (if (and (typep subnode '(or node-application
+                                                                        node-direct-application))
+                                                    (eq 'coalton:Cons (node-rator-name subnode))
```

Comment:
@eliaslfox was right - preventing inlining `Cons` was the only thing necessary to get the recursive data working, or at least to get the tests to all pass.

Reply:
@eliaslfox 🎉

Reply:
@amorphedstar Can we please abstract this test out and write a nice comment?

It's such unusual "business logic" that it's worth not muddying the code too much *and* properly documenting it.

Reply:
here: https://github.com/coalton-lang/coalton/pull/1193/files?diff=unified&w=0#r1707561858